### PR TITLE
Dynamically decide robot purposes/soccer positions

### PR DESCRIPTION
**Pull request by ysims**
This dynamically determines whether a robot should be the defender or striker. It uses a decentralised system where robots decide what position to be in and tell everyone else what their position is.

### CODE REVIEW

File: module/purpose/Soccer/src/Soccer.cpp [hunk 1]
```diff
@@ -146,6 +201,9 @@ namespace module::purpose {
             emit<Task>(std::unique_ptr<FindPurpose>(nullptr));
             emit(std::make_unique<Stability>(Stability::UNKNOWN));
             emit<Task>(std::make_unique<Walk>(Eigen::Vector3d::Zero()));
+            // Set robot to inactive
+            robots[player_id - 1].active   = false;
+            robots[player_id - 1].position = Position::DYNAMIC;
             log<NUClear::INFO>("Idle mode enabled");
         });
```
Comment (Tom0Brien) [inline module/purpose/Soccer/src/Soccer.cpp:209]: Should we set the robot to inactive if in IDLE mode?

**Review by A-Herfel — DISMISSED** [review]
Unable to break this. Tested in NUsight and Webots.

- Started and stopped in penalisation/robot died, striker and defender
- Same cases with a goalie added
- Looked over code

Looks great :D

**Comment by JohanneMontano** [comment]
> Unable to break this. Tested in NUsight and Webots.
> 
> * Started and stopped in penalisation/robot died, striker and defender
> * Same cases with a goalie added
> * Looked over code
> 
> Looks great :D

Since you have a lot of commits to this pr, i reckon its best to wait for someone to thoroughly review it first. 5000 lines will take a bit of time to review, I'll try to go as fast as I can

**Comment by ysims** [comment]
> > Unable to break this. Tested in NUsight and Webots.
> > 
> > * Started and stopped in penalisation/robot died, striker and defender
> > * Same cases with a goalie added
> > * Looked over code
> > 
> > Looks great :D
> 
> Since you have a lot of commits to this pr, i reckon its best to wait for someone to thoroughly review it first. 5000 lines will take a bit of time to review, I'll try to go as fast as I can

The 5000 lines is from a font being added :laughing:

File: module/purpose/Soccer/data/config/nugus3/Soccer.yaml [hunk 2]
```diff
@@ -1,9 +1,3 @@
...
-log_level: INFO
-
-# For the robot playing without listening to GameController
-force_playing: false
-
 # The position of the robot
-# GOALIE, STRIKER, DEFENDER
-position: DEFENDER
+# GOALIE, STRIKER, DEFENDER, DYNAMIC
+position: GOALIE
```
[the remark this answers is not in the document]
Comment (ysims) [reply module/purpose/Soccer/data/config/nugus3/Soccer.yaml:3]: @Tom0Brien prefer to have goalie in one of the config so we most likely will have a robot as goalie, or set to dynamic so if we only have nugus3 it'll be a striker?

**Review by Tom0Brien — APPROVED** [review]

**Review by A-Herfel — APPROVED** [review]

**Pull request merged**